Reformatting and addition of pydocstyle run in CI

octonawish-akcodes opened the pull request:
This closes  #277 
Needed reviews @spbnick 

The following changes added the Reformatting in the whole kcidb repo and followed the conventions stated. 
I also added a pydocstyle CI job in the workflows.

spbnick commented on .github/workflows/test.yml:
```diff
@@ -60,6 +60,8 @@ jobs:
         run: "flake8 kcidb *.py"
       - name: Check python sources with pylint
         run: "pylint kcidb *.py"
+      - name: Check python sources with pycodestyle
```
The name is still referring to `pycodestyle`, not `pydocstyle`. Please fix.

spbnick commented on conftest.py:
```diff
@@ -8,9 +8,9 @@
-))
+CLEAN_DATABASE_SPECS = list(
+    kcidb.db.misc.parse_spec_list(os.environ.get("KCIDB_CLEAN_TEST_DATABASES", ""))
+)
```
Hmm, does `pydocstyle` actually request **code** changes? If so, please explain that in the commit message. If not, and you think this should be done for some other reason, please move this and all other code changes to a separate commit, and explain the purpose and reason for the change in the commit message.

octonawish-akcodes replied:
No, it doesnt. Actually the large amount changes are of the Reformatting done by me as an extra to make it look more good and readable using a formatting extension in vs code.
Should I move all these formatting changes to separate commit or raise a separate issue reg it and then make a PR for that?

spbnick replied:
I appreciate the effort to make the code look nicer. However, please keep **all** reformatting of otherwise-unchanged code in separate commits, and explain why you're doing them in the commit messages. Then we can discuss them. Otherwise they make reviewing harder. 

Please also note that some of your changes take lines over the 79-column limit, and so shouldn't be done, at least in that shape.

Thank you :heart:

octonawish-akcodes replied:
> I appreciate the effort to make the code look nicer. However, please keep **all** reformatting of otherwise-unchanged code in separate commits, and explain why you're doing them in the commit messages. Then we can discuss them. Otherwise, they make reviewing harder.
> 
> Please also note that some of your changes take lines over the 79-column limit, and so shouldn't be done, at least in that shape.
> 
> Thank you heart

Yes they are, actually the thing is I'm a bit lost cuz if I try to maintain them under the 78 limit then the pydoc creates the issue like the sentence should end with '.' not with other alphabets and then sometimes gets an error message of blank space needed after first-line description. So this is happening.

spbnick replied:
Let's treat the docstrings and the code separately. This PR should only concern itself with docstrings, since it's adding the `pydocstyle` to CI. We can discuss and deal with code formatting separately.

Pydocstyle demands that a multi-line docstring should have a summary line, and it **must** be a single line. Currently the code often has first paragraphs (or several) just containing the description of the function/class, and there is **no** summary line as such. Some (most?) of those are short enough to be on a single line, and pydocstyle would count them as "summaries". Some of the others are longer. So, if you want to do this change, you have to come up with a short summary line for those, and put it above the existing description. However, consider these rules, and use your wits to do what makes best sense, it might not always work out like this. This is a documentation-writing exercise to a considerable extent, not only code and text manipulation.

Lastly, read the [PEP 257](https://peps.python.org/pep-0257/) to understand the complete requirements pydocstyle is trying to enforce.

If you'd like we could have a video call about this and other issues, e.g. tomorrow. Just tell me which times you're available, on Slack, and I'll try to find 30-60 minutes.

octonawish-akcodes replied:
Okay

spbnick commented on kcidb/__init__.py:
```diff
@@ -1,29 +1,30 @@
     def __init__(self, database=None, project_id=None, topic_name=None):
-        """
-        Initialize a reporting client
+        """Initialize a reporting client.
```
Hmm, what was the `pydocstyle` error message that prompted this and other similar changes?

I don't particularly like it, as it makes the description harder to notice in a multi-line docstring. Perhaps there's another way?

octonawish-akcodes replied:
In particular I get these kinds of error 

```
/home/abhishek/Downloads/kernelci/kcidb/kcidb/__init__.py:17 in public method `__init__`:
        D107: Missing docstring in __init__
/home/abhishek/Downloads/kernelci/kcidb/kcidb/__init__.py:22 in public class `Client`:
        D400: First line should end with a period (not 't')

```

spbnick replied:
Ah, I meant putting the first line on the same line with the opening `"""`.

It seems it's not necessary (from https://peps.python.org/pep-0257/#multi-line-docstrings):
>  The summary line may be on the same line as the opening quotes or on the next line.

So could you please keep the summary line on the next line, after the opening `"""`?

octonawish-akcodes replied:
Then it throws an error single line comments should end in one line ://

spbnick replied:
Ah, forgot to mention that. Yes. A single-line docstring is fine to be on the same line with **both** the opening **and** closing `"""`.

However, when you have multi-line ones, you tend to miss the first line.

I.e. we want this:
```python
"""Single-line docstring."""
```
and this:
```python
"""
Multi-line docstring summary.

Multi-line docstring elaboration, if any.

Args:
    x: An argument
"""

octonawish-akcodes replied:
Okay

spbnick commented on kcidb/db/__init__.py:
```diff
@@ -8,32 +8,39 @@
-        control.
+        """Retrieve a dictionary of driver names and types available for
+        
+        driver's control.
```
Uh-oh, I don't think this would do. The first line is [supposed](https://peps.python.org/pep-0257/#multi-line-docstrings) to be a summary and the lines after it and the blank line - a more elaborate description. Just spilling the original summary over to the elaborate description misses the point. The right solution here should be at least shortening the original summary until it fits the new limit, and then posting the original summary as the elaborate description. Still nicer would be elaborating more there, but that's optional.

Same goes for similar solutions everywhere else.

octonawish-akcodes replied:
Actually here the pydodstyle was giving me error for adding a blank link after each description first line that's why I have to add it too and I also didn't liked it much.

spbnick reviewed (changes requested):
Wooow, that's a lot of changes and work done! Thank you, Abhishek!

However, I thought pydocstyle wouldn't need **any** code changes. Could you explain why you did them? Please see the inline comments as well.

Also, please get rid of the merge commit in your PR.

Thank you :heart:

octonawish-akcodes commented:
@spbnick I have done all the changes needed for `pydocstyle` and added a run of it too in the CI and have made changes the pdocstyle was giving me in the project and I have fixed them all. Need reviews.

octonawish-akcodes commented:
![image](https://user-images.githubusercontent.com/76171953/232965848-292816e6-a0dc-4b91-80b8-7e846113eb56.png)

What to do now sir?

spbnick commented:
> What to do now sir?

All development package dependencies are listed in `setup.py`, under `extras_require`. They are installed during the setup of the GitHub Actions workflow. You need to add `pydocstyle` there, so it would get installed.

Regarding changes, please make sure there are no merge commits or unrelated changes (like adding the `.github/workflows/script` file) in there, and also rebase your changes onto latest `main`.

I would recommend cloning another copy of the repo and experimenting with cleaning this up there, so you don't lose your changes.

octonawish-akcodes commented:
@spbnick I have done all the necessary changes, Pls review
